src/Nervos/Address.cpp

```diff
@@ -0,0 +1,160 @@
+using namespace TW;
+
+bool Address::isValid(const std::string& string) {
+    return Address::isValid(string, TW::p2pkhPrefix(TWCoinTypeNervos),
+                            TW::p2shPrefix(TWCoinTypeNervos),
+                            stringForHRP(TW::hrp(TWCoinTypeNervos)));
+}
+
+bool Address::isValid(const std::string& string, byte p2pkh, byte p2sh, const char* hrp) {
+    try {
+        Address(string, p2pkh, p2sh, hrp);
+        return true;
+    } catch (std::exception e) {
+        return false;
+    }
+}
+
+Address::Address(const std::string& string)
+    : Address(string, TW::p2pkhPrefix(TWCoinTypeNervos), TW::p2shPrefix(TWCoinTypeNervos),
+              stringForHRP(TW::hrp(TWCoinTypeNervos))) {}
+
+Address::Address(const std::string& string, byte p2pkh, byte p2sh, const char* hrp) : hrp(hrp) {
+    bool isValid = false;
+
+    do {
+        auto decoded = Bech32::decode(string);
+        auto& decodedHrp = std::get<0>(decoded);
+        auto& decodedData = std::get<1>(decoded);
+        auto& decodedVariant = std::get<2>(decoded);
+        if (decodedHrp.compare(hrp)) {
+            break;
+        }
+        Data decodedPayload;
+        if (!Bech32::convertBits<5, 8, false>(decodedPayload, decodedData)) {
+            break;
+        }
+        if (decodedPayload.size() == 0) {
+            break;
+        }
+        addressType = (AddressType)decodedPayload[0];
+        switch (addressType) {
+        case AddressType::FullVersion: {
+            if (decodedVariant != Bech32::ChecksumVariant::Bech32M) {
+                break;
+            }
+            if (decodedPayload.size() < 34) {
+                break;
+            }
+            codeHashIndex = -1;
+            codeHash = Data(decodedPayload.begin() + 1, decodedPayload.begin() + 33);
+            hashType = (HashType)decodedPayload[33];
+            args = Data(decodedPayload.begin() + 34, decodedPayload.end());
+            isValid = true;
+            break;
+        }
+        case AddressType::HashIdx: {
+            if (decodedVariant != Bech32::ChecksumVariant::Bech32) {
+                break;
+            }
+            if (decodedPayload.size() != 22) {
+                break;
+            }
+            codeHashIndex = decodedPayload[1];
+            if (codeHashIndex != 0) {
+                break;
+            }
+            codeHash = CommonFunc::getSecp256k1CodeHash();
+            hashType = HashType::Type1;
+            args = Data(decodedPayload.begin() + 2, decodedPayload.end());
+            isValid = true;
+            break;
+        }
+        case AddressType::DataCodeHash:
+        case AddressType::TypeCodeHash: {
+            if (decodedVariant != Bech32::ChecksumVariant::Bech32) {
+                break;
+            }
+            if (decodedPayload.size() < 33) {
+                break;
+            }
+            codeHashIndex = -1;
+            codeHash = Data(decodedPayload.begin() + 1, decodedPayload.begin() + 33);
```

Comment:
this code is the same as line 64, consider re-writing a function that will fill the member in this case to respect https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

Reply:
Only this 2 lines of code is common for 2 types of addresses (that too just coincidentally). Writing a function for 2 lines of code and passing 5 parameters (some by value, some by reference) might be more inefficient. Still, if you want it, we will do it. You have to right to say and decide in this repo!